src/bun.js/net.exports.js

```diff
@@ -53,9 +53,246 @@ export function isIP(s) {
+            callback(error);
+          }
+          self.emit("error", error);
+        },
+        open(socket) {
+          socket.timeout(self.timeout);
+          self.#socket = socket;
+          self.connecting = false;
+          self.emit("connect");
+          drain(socket);
+        },
+        timeout() {
+          self.emit("timeout");
+        },
+      },
+    });
+    return this;
+
+    function close() {
+      if (ended) return;
+      ended = true;
+      if (self.#readQueue.length > 0) {
+        self.#readQueue.push(null);
+      } else {
+        self.push(null);
+      }
+    }
+
+    function drain(socket) {
+      const { chunk, callback } = self.#writeQueue || {};
+      if (callback) {
+        const written = socket.write(chunk);
+        self.bytesWritten += written;
+        if (written < chunk.length) {
+          self.#writeQueue.chunk = chunk.slice(written);
+        } else {
+          self.#writeQueue = null;
+          callback(null);
+        }
+      }
+    }
+  }
+
+  _destroy(err, callback) {
+    this.#socket?.shutdown();
+    callback(err);
+  }
+
+  _final(callback) {
+    this.#socket.end();
+    callback();
+  }
+
+  get localAddress() {
+    return "127.0.0.1";
+  }
+
+  get localFamily() {
+    return "IPv4";
+  }
+
+  get localPort() {
+    return this.#socket?.localPort;
+  }
+
+  get pending() {
+    return this.connecting;
+  }
+
+  _read(size) {
+    const buffers = this.#readQueue;
+    const len = buffers.length;
+    if (len == 0) return;
+    for (var i = 0; i < len; i++) {
+      if (!this.push(buffers[i])) break;
+    }
+    if (i == len) {
+      buffers.length = 0;
+    } else {
+      this.#readQueue = buffers.slice(i);
+    }
+  }
+
+  get readyState() {
+    if (this.connecting) return "opening";
+    if (this.readable) {
+      return this.writable ? "open" : "readOnly";
+    } else {
+      return this.writable ? "writeOnly" : "closed";
+    }
+  }
+
+  ref() {
+    this.#socket?.ref();
+  }
+
+  get remoteAddress() {
+    return this.#socket.remoteAddress;
+  }
+
+  get remoteFamily() {
+    return "IPv4";
+  }
+
+  resetAndDestroy() {
+    this.#socket?.shutdown();
+  }
+
+  setKeepAlive(enable = false, initialDelay = 0) {
+    // TODO
+  }
+
+  setNoDelay(noDelay = true) {
+    // TODO
+  }
+
+  setTimeout(timeout, callback) {
+    this.#socket?.timeout(timeout);
+    this.timeout = timeout;
+    if (callback) this.once("timeout", callback);
+    return this;
+  }
+
+  unref() {
+    this.#socket?.unref();
+  }
+
+  _write(chunk, encoding, callback) {
+    if (typeof chunk == "string" && encoding !== "utf8") chunk = Buffer.from(chunk, encoding);
+    var written = this.#socket?.write(chunk);
+    if (written == chunk.length) {
+      callback();
+    } else if (this.#writeQueue) {
+      callback(new Error("overlapping _write()"));
+    } else {
+      if (written > 0) chunk = chunk.slice(written);
+      this.#writeQueue = { chunk, callback };
```

Comment:
This will potentially be a lot of prototype transitions.

Instead, could we have this be three private properties on the object?

Reply:
Three? :thinking:

`#writeCallback`, `#writeChunk` and...?

Reply:
A boolean saying "is anything pending?" 

I guess #writeCallback would be undefined and we could use that instead.

Reply:
reused truthiness with two fields, done in https://github.com/oven-sh/bun/pull/1701/commits/81b7104e6517d030ce8aea87609f4b938509a1c3